Add documentation for pod-overrides and container-overrides attributes (#70)
devfile/devfile-web: libs/docs/src/docs/2.2.1-alpha/overriding-pod-and-container-attributes.md

---
title: Extending kubernetes resources
description: Extending kubernetes resources
---

This section describes how you can extend Kubernetes resources for container components through `pod-overrides` and `container-overrides` attributes.

## container-overrides

> `container-overrides` attributes allow you to override properties of a container in a pod spec such as `securityContext` and `resources`. However, it restricts from overriding properties such as `image`, `name`, `ports`, `env`, `volumesMounts`, `command`, and `args`.

Reviewer: Have other implementations, like in DevWorkspace Operator, the same restrictions?

If not, than it needs to be mentioned

Author: At the time of implementation, DWO had the same restriction. I'll check again and update the doc.

    ```

The data inside `container-overrides` can also be specified as a JSON.
```yaml
  container-overrides:
    securityContext: {"runAsUser": "1001", "runAsGroup":"1001"}
```

## pod-overrides

`pod-overrides` attribute allow you to override properties of a pod spec such as `securityContext`, `serviceAccountName`, `schedulerName`, etc. However, it restricts overriding properties such as `containers`, `initContainers`, and `volumes`.

This attribute can be defined at the component and devfile attributes levels. If it is defined at both the levels, the attribute at the devfile attributes level will be parsed first and then the component level.[Strategic Merge Patch](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-api-machinery/strategic-merge-patch.md#basic-patch-format) strategy is used to merge the attributes defined at both the levels.

### Procedure
1. Specify "pod-overrides" at the component level.
    ```yaml {% title="Specify pod-overrides to override security context for container at component level" filename="devfile.yaml" %}
    schemaVersion: 2.2.0
    components:
      - name: maven
        attributes:

>           pod-overrides:

Reviewer: I wanted to test this, but it doesn't seem to work :-(

Author: Can you share the Devfile if you still have it?

Author: You're right. I've created a fix https://github.com/devfile/library/pull/162.

---

> title: Extend Kubernetes resources for container components

Reviewer 2: ```suggestion
title: Extending kubernetes resources
```

> description: Extend Kubernetes resources for container components through attributes

Reviewer 2: ```suggestion
description: Extending kubernetes resources
```

We can also rename this file to `extending-kubernetes-resources` since we don't have a section just for kubernetes resources

---

> This section describes how you can extend kubernetes resources for container components through `pod-overrides` and `container-overrides` attributes.

Reviewer 2: ```suggestion
This section describes how you can extend Kubernetes resources for container components through `pod-overrides` and `container-overrides` attributes.
```

> These attributes can be defined at Devfile level and component level. [Strategic Merge Patch](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-api-machinery/strategic-merge-patch.md#basic-patch-format) strategy is used to merge the attributes defined at both level.

Reviewer 2: ```suggestion
These attributes can be defined at the devfile level and component level. [Strategic Merge Patch](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-api-machinery/strategic-merge-patch.md#basic-patch-format) strategy is used to merge the attributes defined at both levels.
```

## container-overrides

> `container-overrides` attributes allows you to override properties of a container in a pod spec such as `securityContext` and `resources`. However, it restricts from overriding properties such as `image`, `name`, `ports`, `env`, `volumesMounts`, `command`, and `args`.

Reviewer 2: ```suggestion
`container-overrides` attributes allow you to override properties of a container in a pod spec such as `securityContext` and `resources`. However, it restricts from overriding properties such as `image`, `name`, `ports`, `env`, `volumesMounts`, `command`, and `args`.
```

### Procedure

> 1. Specify "container-overrides" at component level.

Reviewer 2: ```suggestion
1. Specify "container-overrides" at the component level.
```

    ```yaml {% title="Specify container-overrides to override security context for container at component level" filename="devfile.yaml" %}

>     ----

Reviewer 2: We can remove the dashes and also add `schemaVersion: <version>` for the devfile examples

Author: What will be the schemaVersion here? 2.2.1 or 2.2.0? Also, do I need to backport this doc to 2.2.0 and 2.1.0?

Author: > We can remove the dashes
All the dashes? Top and bottom ones?

    metadata:
      name: java-quarkus
    ...
    components:
      - name: maven
        attributes:
          container-overrides:
            securityContext:
              runAsUser: 1001
              runAsGroup: 1001
        container:
          image: eclipse/maven-jdk8:latest
    ----
    ```

> 2. Specify "container-overrides" at Devfile level in `metadata` object.

Reviewer 2: ```suggestion
2. Specify "container-overrides" at the devfile level in `metadata` object.
```

    ```yaml {% title="Specify container-overrides to override resources for container at Devfile level" filename="devfile.yaml" %}
    ----
    metadata:
      name: java-quarkus
      attributes:
        container-overrides:
          resources:
            limits:
              memory: 4Gi
              nvidia.com/gpu: 1 # limiting to 1 GPU
            requests:
              nvidia.com/gpu: 1 # requesting 1 GPU
    ...
    components:
      - name: maven
        container:
          image: eclipse/maven-jdk8:latest
    ----
    ```

## pod-overrides

> `pod-overrides` attributes allows you to override properties of a pod spec such as `securityContext`, `serviceAccountName`, `schedulerName`, etc. However, it restricts from overriding properties such as `containers`, `initContainers`, and `volumes`.

Reviewer 2: ```suggestion
`pod-overrides` attributes allow you to override properties of a pod spec such as `securityContext`, `serviceAccountName`, `schedulerName`, etc. However, it restricts overriding properties such as `containers`, `initContainers`, and `volumes`.
```

> If the attributes are defined at both Devfile and component level, the attribute at Devfile level will be parsed first and then the component level.

Reviewer 2: ```suggestion
If the attributes are defined at both devfile and component levels, the attribute at the devfile level will be parsed first and then the component level.
```

### Procedure

> 1. Specify "pod-overrides" at component level.

Reviewer 2: ```suggestion
1. Specify "pod-overrides" at the component level.
```

    ```yaml {% title="Specify container-overrides to override security context for container at component level" filename="devfile.yaml" %}
    ----
    metadata:
      name: java-quarkus
    ...
    components:
      - name: maven
        attributes:
          pod-overrides:
            serviceAccountName: new-service-account
        container:
          image: eclipse/maven-jdk8:latest
    ----
    ```

> 2. Specify "pod-overrides" at Devfile level in `metadata` object.

Reviewer 2: ```suggestion
2. Specify "pod-overrides" at the devfile level in the `metadata` object.
```
